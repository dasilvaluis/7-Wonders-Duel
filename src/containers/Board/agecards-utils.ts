import { flattenMultiLevelArray, centerHorizontally, centerRow, getRowOf } from "./utils";
import { Position, GameElement } from "../../types";
import { CARD_WIDTH, CARD_MARGIN, CARD_HEIGHT } from "../../contants";

export const schemeFirstAge = [ 2, 3, 4, 5, 6 ];
export const schemeSecondAge = [ 6, 5, 4, 3, 2 ];
export const schemeThirdAge = [ 2, 3, 4, 2, 4, 3, 2 ];

const moveRowVertically = (row: Array<Position>, rowIndex: number) => row.map((position) => ({
  ...position,
  y: rowIndex * (CARD_HEIGHT / 3)
}));

export const getAgeScheme = (age: 'I' | 'II' | 'III') => {
  switch (age) {
    case "III":
      return schemeThirdAge;
    case "II":
      return schemeSecondAge;
    default:
      return schemeFirstAge;
  }
}

export const fixThirdAgeCards = (cards: Array<GameElement>) => {
  const _cards = [ ...cards ];
  _cards[9].x = cards[9].x - (CARD_WIDTH + CARD_MARGIN) / 2;
  _cards[10].x = cards[10].x + (CARD_WIDTH + CARD_MARGIN) / 2;

  return _cards;
}

export const getAgeCardsPlacement = (scheme: Array<number>) => {
  const rows = scheme.map((cardsQuantity, rowIndex) => {
    const row = getRowOf(cardsQuantity, CARD_WIDTH);
    const rowMovedVertically = moveRowVertically(row, rowIndex);
    const rowCentered = centerRow(rowMovedVertically, cardsQuantity, CARD_WIDTH);

    return rowCentered;
  });

  return centerHorizontally(flattenMultiLevelArray<Position>(rows));
};