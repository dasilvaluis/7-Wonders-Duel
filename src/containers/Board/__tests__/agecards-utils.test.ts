import { getAgeScheme, getAgeCardsPlacement } from "../agecards-utils"

describe('getAgeCardsPlacement', () => {
  it('places first age cards', () => {
    expect(getAgeCardsPlacement(getAgeScheme('I'))).toMatchSnapshot();
  });

  it('places second age cards', () => {
    expect(getAgeCardsPlacement(getAgeScheme('II'))).toMatchSnapshot();
  });

  it('places third age cards', () => {
    expect(getAgeCardsPlacement(getAgeScheme('III'))).toMatchSnapshot();
  });
});
