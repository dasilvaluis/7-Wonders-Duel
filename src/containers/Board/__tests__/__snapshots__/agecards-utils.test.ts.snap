// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAgeCardsPlacement places first age cards 1`] = `
Array [
  Object {
    "x": 705,
    "y": 0,
  },
  Object {
    "x": 850,
    "y": 0,
  },
  Object {
    "x": 632.5,
    "y": 62.5,
  },
  Object {
    "x": 777.5,
    "y": 62.5,
  },
  Object {
    "x": 922.5,
    "y": 62.5,
  },
  Object {
    "x": 560,
    "y": 125,
  },
  Object {
    "x": 705,
    "y": 125,
  },
  Object {
    "x": 850,
    "y": 125,
  },
  Object {
    "x": 995,
    "y": 125,
  },
  Object {
    "x": 487.5,
    "y": 187.5,
  },
  Object {
    "x": 632.5,
    "y": 187.5,
  },
  Object {
    "x": 777.5,
    "y": 187.5,
  },
  Object {
    "x": 922.5,
    "y": 187.5,
  },
  Object {
    "x": 1067.5,
    "y": 187.5,
  },
  Object {
    "x": 415,
    "y": 250,
  },
  Object {
    "x": 560,
    "y": 250,
  },
  Object {
    "x": 705,
    "y": 250,
  },
  Object {
    "x": 850,
    "y": 250,
  },
  Object {
    "x": 995,
    "y": 250,
  },
  Object {
    "x": 1140,
    "y": 250,
  },
]
`;

exports[`getAgeCardsPlacement places second age cards 1`] = `
Array [
  Object {
    "x": 415,
    "y": 0,
  },
  Object {
    "x": 560,
    "y": 0,
  },
  Object {
    "x": 705,
    "y": 0,
  },
  Object {
    "x": 850,
    "y": 0,
  },
  Object {
    "x": 995,
    "y": 0,
  },
  Object {
    "x": 1140,
    "y": 0,
  },
  Object {
    "x": 487.5,
    "y": 62.5,
  },
  Object {
    "x": 632.5,
    "y": 62.5,
  },
  Object {
    "x": 777.5,
    "y": 62.5,
  },
  Object {
    "x": 922.5,
    "y": 62.5,
  },
  Object {
    "x": 1067.5,
    "y": 62.5,
  },
  Object {
    "x": 560,
    "y": 125,
  },
  Object {
    "x": 705,
    "y": 125,
  },
  Object {
    "x": 850,
    "y": 125,
  },
  Object {
    "x": 995,
    "y": 125,
  },
  Object {
    "x": 632.5,
    "y": 187.5,
  },
  Object {
    "x": 777.5,
    "y": 187.5,
  },
  Object {
    "x": 922.5,
    "y": 187.5,
  },
  Object {
    "x": 705,
    "y": 250,
  },
  Object {
    "x": 850,
    "y": 250,
  },
]
`;

exports[`getAgeCardsPlacement places third age cards 1`] = `
Array [
  Object {
    "x": 705,
    "y": 0,
  },
  Object {
    "x": 850,
    "y": 0,
  },
  Object {
    "x": 632.5,
    "y": 62.5,
  },
  Object {
    "x": 777.5,
    "y": 62.5,
  },
  Object {
    "x": 922.5,
    "y": 62.5,
  },
  Object {
    "x": 560,
    "y": 125,
  },
  Object {
    "x": 705,
    "y": 125,
  },
  Object {
    "x": 850,
    "y": 125,
  },
  Object {
    "x": 995,
    "y": 125,
  },
  Object {
    "x": 705,
    "y": 187.5,
  },
  Object {
    "x": 850,
    "y": 187.5,
  },
  Object {
    "x": 560,
    "y": 250,
  },
  Object {
    "x": 705,
    "y": 250,
  },
  Object {
    "x": 850,
    "y": 250,
  },
  Object {
    "x": 995,
    "y": 250,
  },
  Object {
    "x": 632.5,
    "y": 312.5,
  },
  Object {
    "x": 777.5,
    "y": 312.5,
  },
  Object {
    "x": 922.5,
    "y": 312.5,
  },
  Object {
    "x": 705,
    "y": 375,
  },
  Object {
    "x": 850,
    "y": 375,
  },
]
`;
